packages/peregrine/lib/talons/CartPage/PriceAdjustments/GiftOptions/useGiftOptions.js

```diff
@@ -16,145 +20,360 @@ import DEFAULT_OPERATIONS from './giftOptions.gql';
+export const useGiftOptions = (props = {}) => {
     const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
-    const { getGiftOptionsQuery } = operations;
+    const { setGiftOptionsOnCartMutation, getGiftOptionsQuery } = operations;
 
     const [{ cartId }] = useCartContext();
-    const { cache } = useApolloClient();
-
-    // The form will be a child, not a parent, so we can't call `useFormApi`.
-    // Need to use a ref and `props.getApi` instead.
-    const formApiRef = useRef();
-    const [hasHydrated, setHasHydrated] = useState(false);
-
-    // Could be an effect, but a callback should be slightly better.
-    const handleCompleted = useCallback(
-        data => {
-            // Only write values to the form once, ideally before user input.
-            // Afterward, treat client state as the single source of truth.
-            if (data && !hasHydrated) {
-                formApiRef.current.setValues({
-                    cardMessage: data.cart.local_gift_message,
-                    includeGiftReceipt: data.cart.include_gift_receipt,
-                    includePrintedCard: data.cart.include_printed_card
+
+    const [
+        setGiftOptionsOnCart,
+        { error: setGiftOptionsOnCartError }
+    ] = useMutation(setGiftOptionsOnCartMutation);
+    const {
+        data: getGiftOptionsData,
+        error: getGiftOptionsError,
+        loading
+    } = useQuery(getGiftOptionsQuery, {
+        fetchPolicy: 'cache-and-network',
+        nextFetchPolicy: 'cache-first',
+        skip: !cartId,
+        variables: { cartId }
+    });
+
+    const { cart } = getGiftOptionsData || {};
+
+    const formApiRef = useRef(null);
+
+    const [giftMessageIsChecked, setGiftMessageIsChecked] = useState(false);
+    const [showGiftMessageResult, setShowGiftMessageResult] = useState(false);
+    const [savingOptions, setSavingOptions] = useState([]);
+
+    const initialValues = useMemo(() => {
+        const hasInitialGiftMessage =
+            cart?.gift_message?.from.length > 0 ||
+            cart?.gift_message?.to.length > 0 ||
+            cart?.gift_message?.message.length > 0;
+
+        setGiftMessageIsChecked(hasInitialGiftMessage);
+        setShowGiftMessageResult(hasInitialGiftMessage);
+
+        return {
+            cardFrom: cart?.gift_message?.from || '',
+            cardTo: cart?.gift_message?.to || '',
+            cardMessage: cart?.gift_message?.message || '',
+            includeGiftReceipt: cart?.gift_receipt_included === true,
+            includeGiftMessage: hasInitialGiftMessage,
+            includePrintedCard: cart?.printed_card_included === true
+        };
+    }, [cart]);
+
+    const giftMessageResult = useMemo(
+        () => ({
+            cardFrom: cart?.gift_message?.from || '',
+            cardTo: cart?.gift_message?.to || '',
+            cardMessage: cart?.gift_message?.message || ''
+        }),
+        [
+            cart?.gift_message?.from,
```

Comment:
Can't we just use cart.gift_message as dependencies here?

Reply:
That's what the linter gave me.
From my understanding, if we only use `cart.gift_message`, it won't update if only one of the object entry is updated.
Because it needs to listen to each entry.

Reply:
I see, ok.